File: C++/stack.cpp
#include "stack.h"

#include <istream>
#include <ostream>
#include <utility>
#include <vector>

namespace {
void write_string_bin(std::ostream& os, const std::string& s) {
    const std::uint32_t n = static_cast<std::uint32_t>(s.size());
    os.write(reinterpret_cast<const char*>(&n), sizeof(n));
    os.write(s.data(), static_cast<std::streamsize>(n));
}
std::string read_string_bin(std::istream& is) {
    std::uint32_t n = 0;
    is.read(reinterpret_cast<char*>(&n), sizeof(n));
    std::string s(n, '\0');
    if (n > 0) {
        is.read(&s[0], static_cast<std::streamsize>(n));
    }
    return s;
}
}

Stack::Stack(std::string name) : name(std::move(name)) {}

Stack::~Stack() { clear(); }

void Stack::clear() {
    Node* cur = top_;
    while (cur) {
        Node* nxt = cur->next;
        delete cur;
        cur = nxt;
    }
    top_ = nullptr;
}

void Stack::push(const std::string& data) {
    auto* n = new Node(data);
    n->next = top_;
    top_ = n;
}

std::string Stack::pop() {
    if (!top_) {
        return "";
    }
    Node* del = top_;
    std::string out = del->data;
    top_ = top_->next;
    delete del;
    return out;
}

std::string Stack::peek() const { return top_ ? top_->data : ""; }

bool Stack::is_empty() const { return top_ == nullptr; }

std::string Stack::read() const {
    std::string out;
    bool first = true;
    for (Node* cur = top_; cur; cur = cur->next) {
        if (!first) {
            out += ' ';
        }
        first = false;
        out += cur->data;
    }
    return out;
}

void Stack::serialize_binary(std::ostream& os) const {
    std::vector<std::string> values;
    for (Node* cur = top_; cur; cur = cur->next) {
        values.push_back(cur->data);
    }
    const std::uint32_t sz = static_cast<std::uint32_t>(values.size());
    os.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    for (const auto& v : values) {
        write_string_bin(os, v);
    }
}

void Stack::deserialize_binary(std::istream& is) {
    clear();
    std::uint32_t sz = 0;
    is.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    std::vector<std::string> values;
    values.reserve(sz);
    for (std::uint32_t i = 0; i < sz; ++i) {
        values.push_back(read_string_bin(is));
    }
    for (std::size_t i = values.size(); i-- > 0;) {
        push(values[i]);
    }
}

void Stack::serialize_text(std::ostream& os) const {
    std::vector<std::string> values;
    for (Node* cur = top_; cur; cur = cur->next) {
        values.push_back(cur->data);
    }
    os << values.size() << "\n";
    for (const auto& v : values) {
        os << v << "\n";
    }
}

void Stack::deserialize_text(std::istream& is) {
    clear();
    int sz = 0;
    if (!(is >> sz)) {
        return;
    }
    std::string dummy;
    std::getline(is, dummy);
    std::vector<std::string> values;
    values.reserve(static_cast<std::size_t>(sz));
    for (int i = 0; i < sz; ++i) {
        std::string line;
        std::getline(is, line);
        values.push_back(line);
    }
    for (std::size_t i = values.size(); i-- > 0;) {
        push(values[i]);
    }
}

File: datastructures-go/stack.go
package datastructures

import (
	"encoding/binary"
	"fmt"
	"io"
	"strings"
)

type stackNode struct {
	data string
	next *stackNode
}

type Stack struct {
	Name string
	top  *stackNode
}

func NewStack(name string) *Stack {
	return &Stack{Name: name}
}

func (s *Stack) clear() {
	s.top = nil
}

func (s *Stack) Push(data string) {
	node := &stackNode{data: data}
	node.next = s.top
	s.top = node
}

func (s *Stack) Pop() string {
	if s.top == nil {
		return ""
	}
	data := s.top.data
	s.top = s.top.next
	return data
}

func (s *Stack) Peek() string {
	if s.top == nil {
		return ""
	}
	return s.top.data
}

func (s *Stack) IsEmpty() bool {
	return s.top == nil
}

func (s *Stack) Read() string {
	var values []string
	current := s.top
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	return strings.Join(values, " ")
}

func (s *Stack) SerializeBinary(w io.Writer) error {
	var values []string
	current := s.top
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	
	length := uint32(len(values))
	if err := binary.Write(w, binary.LittleEndian, length); err != nil {
		return err
	}
	
	for _, v := range values {
		strLen := uint32(len(v))
		if err := binary.Write(w, binary.LittleEndian, strLen); err != nil {
			return err
		}
		if _, err := w.Write([]byte(v)); err != nil {
			return err
		}
	}
	return nil
}

func (s *Stack) DeserializeBinary(r io.Reader) error {
	s.clear()
	
	var length uint32
	if err := binary.Read(r, binary.LittleEndian, &length); err != nil {
		return err
	}
	
	temp := make([]string, length)
	for i := 0; i < int(length); i++ {
		var strLen uint32
		if err := binary.Read(r, binary.LittleEndian, &strLen); err != nil {
			return err
		}
		
		buf := make([]byte, strLen)
		if _, err := io.ReadFull(r, buf); err != nil {
			return err
		}
		temp[i] = string(buf)
	}
	
	for i := len(temp) - 1; i >= 0; i-- {
		s.Push(temp[i])
	}
	return nil
}

func (s *Stack) SerializeText(w io.Writer) error {
	var values []string
	current := s.top
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	
	if _, err := fmt.Fprintf(w, "%d\n", len(values)); err != nil {
		return err
	}
	
	for _, v := range values {
		if _, err := fmt.Fprintf(w, "%s\n", v); err != nil {
			return err
		}
	}
	return nil
}

func (s *Stack) DeserializeText(r io.Reader) error {
	s.clear()
	
	var length int
	if _, err := fmt.Fscanf(r, "%d\n", &length); err != nil {
		return err
	}
	
	temp := make([]string, length)
	for i := 0; i < length; i++ {
		var data string
		if _, err := fmt.Fscanf(r, "%s\n", &data); err != nil {
			return err
		}
		temp[i] = data
	}
	
	for i := len(temp) - 1; i >= 0; i-- {
		s.Push(temp[i])
	}
	return nil
}
